torch/_inductor/ir.py

```diff
@@ -3009,14 +3009,22 @@ def normalize_start_end(cls, x, dim, start, end):  # type: ignore[no-untyped-def
         dim_size = x.get_size()[dim]
 
         if any(free_unbacked_symbols(x) for x in (start, end, dim_size)):
-
-            def clamp(x, lower, upper):  # type: ignore[no-untyped-def]
-                return sympy.Min(sympy.Max(x, lower), upper)
-
+            min_func = sympy.Min
+            max_func = sympy.Max
         else:
+            min_func = sizevars.evaluate_min
+            max_func = sizevars.evaluate_max
 
-            def clamp(x, lower, upper):  # type: ignore[no-untyped-def]
-                return sizevars.evaluate_min(sizevars.evaluate_max(x, lower), upper)
+        def clamp(x, lower, upper):  # type: ignore[no-untyped-def]
+            clamped_lower = (
+                x if sizevars.statically_known_geq(x, lower) else max_func(x, lower)
+            )
+            clamped_full = (
+                clamped_lower
+                if sizevars.statically_known_leq(clamped_lower, upper)
+                else min_func(clamped_lower, upper)
+            )
+            return clamped_full
```

Comment:
This change is noncontroversial and you might consider landing it separately from the divide by key change which is more controversial.

Reply:
Endorsing this change (though, as a nit, I feel there is some missing abstraction that would be make this less verbose to write at this end).

Reply:
good idea, I lifted this fix out into #143526